Refactor Application.cpp

HifiExperiments opened the pull request:
Closes #931 

(apologies to whoever reviews this)

Application.cpp has gotten totally out of control over the years (in many cases because of my dumb decisions).  This is an attempt to corral it.  I've organized the code as best I could into sections - Assets, Camera, Entities, Events, Graphics, Plugins, Setup, and UI - and divided the implementations into separate, smaller files.  I completely rebuilt the includes from scratch so we only have what's necessary.  I broke some of the classes into their own files.  Overall, I changed very little functionally.  I removed a few unused variables and functions but nearly everything is directly copy-pasted.

Testing-wise, things should just work as before.  In particular, anything related to startup, login UI, and plugins.

## Funding

This project is funded through [NGI0 Entrust](https://nlnet.nl/entrust), a fund established by [NLnet](https://nlnet.nl) with financial support from the European Commission's [Next Generation Internet](https://ngi.eu) program. Learn more at the [NLnet project page](https://nlnet.nl/project/Overte).

[<img src="https://nlnet.nl/logo/banner.png" alt="NLnet foundation logo" width="20%" />](https://nlnet.nl)
[<img src="https://nlnet.nl/image/logos/NGI0_tag.svg" alt="NGI Zero Logo" width="20%" />](https://nlnet.nl/entrust)

HifiExperiments commented on interface/src/Application.h:
```diff
@@ -722,167 +782,124 @@ private slots:
-
-    DiscordPresence* _discordPresence{ nullptr };
 
-    bool _profilingInitialized { false };
```
oh yeah @daleglass it didn't look like this was ever set to true, so I deleted it + the one place it was used, is that ok?

HifiExperiments replied:
it seemed to be working ok for me without it, maybe if it's a problem during testing I can add it back?

daleglass replied:
Sure, it's simple enough that it'd be easy. I did quite a lot of rearranging of that code, and right now can't recall if maybe I just fixed this in some other way in the end.

HifiExperiments commented:
hm

```
/__w/overte/overte/interface/src/scripting/PerformanceScriptingInterface.cpp:14:10: fatal error: QtQML: No such file or directory
   14 | #include <QtQML>
      |          ^~~~~~~
```

I'll look into that...

ksuprynowicz reviewed (changes requested):
This is really awesome :)
Application.cpp needed refactoring badly and this will make future work a lot easier.
I didn't find any problems, but I have few questions and I found what I think is earlier thread safety issue in the code.

ksuprynowicz commented on interface/src/ApplicationMeshProvider.h:
```diff
@@ -0,0 +1,67 @@
+        scriptable::ModelProviderPointer provider;
+        auto entityTreeRenderer = qApp->getEntities();
+        auto entityTree = entityTreeRenderer->getTree();
+        if (auto entity = entityTree->findEntityByID(entityID)) {
```
Maybe as a later PR we should do a thread safety audit of all the places where entity tree is used?

HifiExperiments replied:
oh wow great catch.  `EntityTree::findEntityByID` actually calls `EntityTree::findEntityByEntityItemID` which does internally use a read lock, but then `EntityTreeRenderer::renderableForEntityId` looks like it needs it?  I filed https://github.com/overte-org/overte/issues/1015

ksuprynowicz commented:
I'm also having trouble building it on Linux using Overte-builder:
![image](https://github.com/overte-org/overte/assets/11568651/fcfe0b20-3ea9-4913-8224-2f811efa8ada)

HifiExperiments commented on interface/src/Application.cpp:
```diff
@@ -2835,1134 +1154,381 @@ void Application::onAboutToQuit() {
-        tracer->serialize(outputFile);
-    }
+    // DONT CHECK IN
+    //DependencyManager::get<LODManager>()->setAutomaticLODAdjust(false);
```
done!

HifiExperiments commented on interface/src/Application_Assets.cpp:
```diff
@@ -0,0 +1,525 @@
+        } else {
+            // to prevent files that aren't models or texture files from being loaded into world automatically
+            if ((filePath.toLower().endsWith(OBJ_EXTENSION) || filePath.toLower().endsWith(FBX_EXTENSION)) ||
+                ((filePath.toLower().endsWith(JPG_EXTENSION) || filePath.toLower().endsWith(PNG_EXTENSION)) &&
```
done!

HifiExperiments commented on interface/src/Application_Setup.cpp:
```diff
@@ -0,0 +1,2153 @@
+#endif
+
+using namespace std;
+
```
huh maybe?  how do I do that?

ksuprynowicz commented:
I tested it on Windows, and it asks to log in every time. A different than usual login screen is used for this:
![overte-snap-by--on-2024-06-15_15-10-47](https://github.com/overte-org/overte/assets/11568651/0e0f7b82-17ee-4b17-8abd-681c121d3cf9)
 Kraftwerk world becomes entirely white after skybox loads:
 
![overte-snap-by--on-2024-06-15_15-02-21](https://github.com/overte-org/overte/assets/11568651/5c00dc5f-1e97-4f6b-8494-d20cf2551628)
No audio devices are detected:
![obraz](https://github.com/overte-org/overte/assets/11568651/84a3f726-0dbc-4cef-9e41-e049507d2bac)

ksuprynowicz commented:
It also seems that login screen that it uses in VR is intended for desktop?

ksuprynowicz commented:
The good thing about that different login screen though is that it has option to continue anonymously. Ours is badly missing this.

HifiExperiments commented:
I believe I've fixed the audio device issue.  for the login screen, it turns out I had accidentally deleted `_disableLoginScreen` which was always set to true.  so whatever that fullscreen login page was for, it appears to have been disabled intentionally at some point.  I've re-disabled it to just avoid messing with stuff, but maybe we should look into re-enabling it properly at some point.  I'm not sure if this also fixed the VR login screen, if that's still a problem could you send me a screenshot?

I wasn't able to reproduce the issue in kraftwerk; I wonder if the download got corrupted or something?  does it still happen if you clear your caches, including ktx cache?

JulianGro commented:
I gave this a test on Linux and so far, everything works fine. I tried clearing shader cache and ktx cache as well. In both scenarios, Kraftwerk renders properly.

ksuprynowicz commented:
I tested it again and now everything works well :)

ksuprynowicz commented:
Would someone like to test in VR on Windows? I don't have a VR setup at the moment. If all works well we could merge it.

HifiExperiments commented:
there are some conflicts now, I’ll probably wait until after the next release to rebase and resolve them just to avoid causing any problems

Armored-Dragon commented:
> Would someone like to test in VR on Windows? I don't have a VR setup at the moment. If all works well we could merge it.

I did some brief testing in VR and everything seems to work as expected. 

> there are some conflicts now, I’ll probably wait until after the next release to rebase and resolve them just to avoid causing any problems

I will also take a look at this again after the rebase!

HifiExperiments commented:
I've rebased this and it's all ready to be reviewed + tested again!

HifiExperiments commented:
I've rebased this again and it's still ready for final testing + review!

ksuprynowicz commented:
I just tested it on Linux and on every start no microphone is selected. Default microphone needs to be selected every start. Output seems to be selected properly.

ksuprynowicz commented:
It seems that microphone issue happens only on Linux. It happens on every start and no input is selected:
![image](https://github.com/user-attachments/assets/ed960112-ed32-4eb4-ac65-c0ce676a391b)

ksuprynowicz commented:
It looks like the microphone issue doesn't happen in Desktop mode on Windows, but does happen in VR mode on Windows, exactly like in Desktop mode on Linux.

ksuprynowicz commented:
Teleport script in the tower near Overte_hub spawn point is broken, it fails with message:
```
[01/05 19:47:37] [CRITICAL] [overte.scriptengine] [about:Entities 1] Function call failed: "failed on line 26 column 8 with message: "Uncaught ReferenceError: Window is not defined" backtrace: ReferenceError: Window is not defined
[01/05 19:47:37] [CRITICAL] [overte.scriptengine]     at Object.enterEntity (https://aleziakurdis.github.io/inertia/teleporters/teleporter.js:26:9)
[01/05 19:47:37] [CRITICAL] [overte.scriptengine] [about:Entities 1] JS function call failed:
```

ksuprynowicz commented:
After some more testing input selection issue seems to happen sometimes in Desktop mode on Windows too.

HifiExperiments commented:
darn!  thank you for testing, I'll look into those

HifiExperiments commented:
there’s definitely some weirdness around audio devices on startup.  the previous fix (to the much earlier issue of the audio devices not showing up) was to reorder some initialization: https://github.com/overte-org/overte/pull/939/commits/2345bcc2b58d2cb8560cbabcca65f191ee0e81de#diff-379416a9df19174e2bf0d9d644da8269931fbe2f7f0516719eedb75f8d022b34R810

so…perhaps there’s still something wrong about the order of that versus the loading from settings

HifiExperiments commented:
thanks again to @ksuprynowicz, the audio issue appears to be fixed!

I haven't looked at the teleport issue yet, if that's still an issue I will take a look this weekend, but otherwise I think this is ready for more testing

ksuprynowicz commented:
We did a lot of testing on the newest version of this PR, both in desktop and VR. All seems to work good :)

HifiExperiments commented:
amazing!  this is ready to merge then from my side. 🎉

ksuprynowicz reviewed (approved):
Everything looks good :)

HifiExperiments commented:
HUGE thank you @ksuprynowicz for all your help with this!

ksuprynowicz commented:
@HifiExperiments Thank you for the refactor - it was a massive undertaking and was really needed with how big Application.cpp got XD
